lib/optimize/ModuleConcatenationPlugin.js

```diff
@@ -0,0 +1,209 @@
+						if(module.getNumberOfChunks() !== 1)
+							return false;
+
+						// Because of variable renaming we can't use modules with eval
+						if(module.meta && module.meta.hasEval)
+							return false;
+
+						return true;
+					});
+					const possibleInners = new Set(relevantModules.filter(module => {
+						// Module must not be the entry points
+						if(chunk.entryModule === module)
+							return false;
+
+						// Exports must be known (and not dynamic)
+						if(!Array.isArray(module.providedExports))
+							return false;
+
+						// Using dependency variables is not possible as this wraps the code in a function
+						if(module.variables.length > 0)
+							return false;
+
+						// Module must only be used by Harmony Imports
+						if(!module.reasons.every(reason => reason.dependency instanceof HarmonyImportDependency))
+							return false;
+
+						// It must be statically known which exports are provided or used
+						if(!Array.isArray(module.providedExports))
+							return false;
+
+						return true;
+					}));
+					const possibleRoots = relevantModules.filter(module => {
+						return true;
+					});
+					const concatConfigurations = [];
+					while(possibleRoots.length) {
+						const currentRoot = possibleRoots.pop();
+						// console.log("#process", currentRoot.debugId, currentRoot.resource);
+						const currentConfiguration = new ConcatConfiguration(currentRoot);
+						for(let imp of this.getImports(currentRoot)) {
+							this.tryToAdd(currentConfiguration, imp, possibleInners);
+						}
+						if(!currentConfiguration.isEmpty())
+							concatConfigurations.push(currentConfiguration);
+					}
+					concatConfigurations.sort((a, b) => {
+						return b.modules.size - a.modules.size;
+					});
+					const usedModules = new Set();
+					for(const config of concatConfigurations) {
+						if(usedModules.has(config.rootModule))
+							continue;
+						const orderedModules = new Set();
+						this.addInOrder(config.rootModule, config.modules, orderedModules);
+						const newModule = new ConcatenatedModule(config.rootModule, Array.from(orderedModules));
+						for(const m of orderedModules) {
+							usedModules.add(m);
+							chunk.removeModule(m);
+						}
+						chunk.addModule(newModule);
+						compilation.modules.push(newModule);
+						if(chunk.entryModule === config.rootModule)
+							chunk.entryModule = newModule;
+						config.rootModule.reasons.forEach(reason => {
+							if(!config.modules.has(reason.module))
+								reason.dependency.module = newModule;
+						});
+					}
+				});
+			});
+		});
+	}
+
+	getImports(module) {
+		return Array.from(new Set(module.dependencies
+
+			// Only harmony Dependencies
+			.filter(dep => dep instanceof HarmonyImportDependency && dep.module)
+
+			// Dependencies are simple enough to concat them
+			.filter(dep => {
+				return !module.dependencies.some(d =>
+					d instanceof HarmonyExportImportedSpecifierDependency &&
+					d.importDependency === dep &&
```

Comment:
`export * from "xyz"` is a pair of `HarmonyImportDependency` + `HarmonyExportImportedSpecifierDependency`.
This snippet finds all `HarmonyExportImportedSpecifierDependency` for a `HarmonyImportDependency`.